File: ELEC881_Tabnine/src/No Description Prompt/Python/q30-frog-jump.py
class Solution(object):
    def canCross(self, stones):
        """
        :type stones: List[int]
        :rtype: bool
        """
        if len(stones) < 3:
            return False

        dp = {stone: {1} for stone in stones}
        for stone in stones:
            for k in dp[stone]:
                for next_stone in (stone + k - 1, stone + k, stone + k + 1):
                    if next_stone in stones:
                        dp[next_stone] = dp.get(next_stone, set()) | {k}

        return len(dp[stones[-1]]) > 0
        

File: ELEC881_Tabnine/src/No Description Prompt/Java script (JS)/q30-frog-jump.js
/**
 * @param {number[]} stones
 * @return {boolean}
 */
var canCross = function(stones) {
    const dp = new Map();
    dp.set(stones[0], true);
    for (let i = 1; i < stones.length; i++) {
        const stone = stones[i];
        for (const [prevStone, canCross] of dp) {
            if (canCross >= Math.abs(stone - prevStone - 1) && dp.has(stone - 2 * Math.abs(stone - prevStone - 1))) {
                dp.set(stone, true);
                break;
            }
        }
        if (!dp.has(stone)) {
            return false;
        }
        dp.delete(prevStone);
        dp.set(stone, false);
        if (dp.has(stone - 1)) {
            dp.set(stone - 1, false);
        }
        if (dp.has(stone + 1)) {
            dp.set(stone + 1, false);
        }
    }
    return true;
};